pages/admin/lessons/[lessonSlug]/[pageName]/index.tsx

```diff
@@ -39,26 +53,76 @@ type Module = {
 }
 type Modules = Module[]
 
-type ContentProps = {
-  pageName?: string | string[]
+const IntroductionPage = ({ lesson }: { lesson: Lesson }) => {
+  const [updateLesson] = useUpdateLessonMutation()
+
+  const [formOptions, setFormOptions] = useState(
+    getPropertyArr(lesson, ['challenges', '__typename'])
+  )
+
+  // Update the Inputs values when the lesson changes
+  useEffect(
```

Comment:
why do you need `useEffect` here? Would this render when lesson change automatically?

Reply:
It's possible to switch between lessons from the Breadcrumbs lessons dropdown menu. When the admin switches to a different lesson, the input value remain the same. To solve it, this `useEffect` has been added.

Reply:
I think there is something wrong with the implementation of logic. We should not be having to use useEffect for this use case. A proper re-render when props change should solve the issue

Can you try giving both child and parent components the same key value?

Reply:
> Can you try giving both child and parent components the same key value?

What do you mean?

Reply:
A few possibilities I was considering:

1) I was thinking that since React sometimes relies on keys to decide when to rerender, then possible placing keys could solve the issue, so that is one possible fix. 
Try giving the IntroductionPage component a key value that changes along with lesson value:
```
return <IntroductionPage lesson={lesson} key={lesson.id} />
```

2) Another possibility for the cause of this issue is that maybe there is no state/prop change. What does the breadcrumb component do? Does it change any states? How does it achieve the effect of switching lessons? I see on line 227 we do `router.push`, so does that mean the breadcrumb component achieves the switching of lessons by redirection to a URL for a specific lesson?
https://github.com/garageScript/c0d3-app/pull/2199/files#diff-a754b3c648a6d2067946a27fc92b81619719063cde89810678ba5d9ce444231aR227

Does this mean that

3) Maybe the state change is still referencing the same object(thus same memory address, so react doesn't see it an update in value).

Reply:
The component re-renders when the lesson changes. When it re-renders, the initial state `formOptions` value `getPropertyArr(lesson, ['challenges', '__typename'])` doesn't get set again because the component got re-rendered. 

It only does when the component first mounts (as far as I tested). To workaround this, one solution is to set the state value again through `useEffect`.

If you tried to comment the `useEffect` code, went to the introduction tab, switched to another lesson, went to the modules tab, and finally back to the introduction tab, you'll notice the inputs have the data of the new lesson. 

This is because it got dismounted. The state initial value is now the new lesson.

Reply:
I see, thanks for additional context. My first suggestion  to give IntroductionPage a key prop should work to solve this issue then.

Reply:
> I see, thanks for additional context. My first suggestion to pass IntroductionPage a key prop should work to solve this issue then.

What will this solve? We're passing it so the component re-renders? 

I applied the suggested changes and it still didn't update the state `formOptions` with the new lesson data.

Reply:
Hmm looks like the issue is not to do with passing a key then. And yeah it would have been better. Key is a special react prop used by react to detect when a component has changed. It isn't supposed to be used for any other purpose

Reply:
Actually, `key={lesson.id}` works. I didn't know the `key` prop could be used to reset the component's states. My first attempt was `lessonId={lesson.id}` which didn't work. But I tried now with the key prop and it worked. Super nice 😄